File: 1001~1500/1420. Build Array Where You Can Find The Maximum Exactly K Comparisons/main.py
"""
Source: https://rb.gy/1vycg
Date: 2023/8/19
Skill:
Ref:
Runtime: 1197 ms, faster than 51.61%
Memory Usage: 20.44 MB, less than 66.13%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right
from functools import cache


class Solution:
    def numOfArrays(self, n: int, m: int, k: int) -> int:
        # * dp[i][j][k]表示對於前i個元素、最大值等於j、總共用了k次cost時，總共有多少種方案
        mod, dp, res = 10 ** 9 + 7, [[[0 for _ in range(k + 1)] for _ in range(m)] for _ in range(n)], 0
        for i in range(m):
            dp[0][i][1] = 1
        for i in range(1, n):
            for j in range(m):
                for l in range(1, k + 1):
                    dp[i][j][l] = dp[i - 1][j][l] * (j + 1)
                    for t in range(0, j):
                        dp[i][j][l] += dp[i - 1][t][l - 1]
                    dp[i][j][l] %= mod

        for i in range(k - 1, m):
            res = (res + dp[-1][i][-1]) % mod
        return res


if __name__ == "__main__":
    s = Solution()

File: 1001~1500/1420. Build Array Where You Can Find The Maximum Exactly K Comparisons/main.cpp
/**
 * Source: ibit.ly/ScmAg
 * Date: 2023/10/8
 * Skill:
 * Runtime: 3 ms, faster than 98.64% of C++ online submissions
 * Memory Usage: 7.48 MB, less than 65.00% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
    ll mod = 1e9+7;
public:
    int numOfArrays(int n, int m, int k) {
        ll dp[n][m+1][k+1];
        for (int i=0; i<n; i++)
            for (int j=0; j<=m; j++)
                for (int l=0; l<=k; l++)
                    dp[i][j][l] = 0;
        // vector<vector<vector<ll>>> dp(n, vector<vector<ll>>(m + 1, vector<ll>(k + 1, 0)));
        for (int i = 1; i <= m; i++)
            dp[0][i][1] = i;
        for (int i = 1; i < n; i++) {
            dp[i][1][1] = 1;
            for (int j = 2; j <= m; j++) {
                dp[i][j][1] = (dp[i][j - 1][1] + (dp[i - 1][j][1] - dp[i - 1][j - 1][1]) * j % mod) % mod;
                int u = min(i + 1, j);
                for (int t = 2; t <= min(u, k); t++)
                    dp[i][j][t] = (dp[i][j - 1][t] % mod + dp[i - 1][j - 1][t - 1] % mod + (dp[i - 1][j][t] - dp[i - 1][j - 1][t]) * j % mod) % mod;
            }
        }
        return (dp[n - 1][m][k] + mod) % mod;
    }
};


int main() {
    Solution s;
    vector<int> nums{2, 3, 3, 2};
    vector<vector<int>> arr{{1,2}, {2,3}, {2,4}};
}